### PR TITLE
chore: improve analytics

### PR DESCRIPTION
**Pull request by mistercrunch**
### SUMMARY
Adding a few fields to what's logged:
- path: the path portion of the url ie `/superset/explore_json/`
- path_no_int:  the path portion of the url, replace int portions `/api/v1/chart/<int>/`
- ref: a reference to the Python object ie: `ChartRestApi.post`

Also instrumenting more events in `/superset/views/core.py`, and the dashboard and chart APIs

Screen Shot 2020-11-16 at 8 53 25 AM

### CODE REVIEW

**Comment by mistercrunch** [comment]
@dpgaspar, looking for ways to cover things like `get_list` in `base_api`, but also want to prevent double-logging. Looking for input on this.

**Comment by mistercrunch** [comment]
@dpgaspar this is ready for review!

File: superset/utils/log.py [hunk 1]
```diff
@@ -69,7 +89,8 @@ def log_context(self, action: str) -> Iterator[Callable[..., None]]:
         except (TypeError, ValueError):
             slice_id = 0
 
-        self.stats_logger.incr(action)
+        if log_to_statsd:
+            self.stats_logger.incr(action)
```
Comment (dpgaspar) [inline superset/utils/log.py:93]: +1 to eventually push to completely remove it

File: superset/views/core.py [hunk 2]
```diff
@@ -731,6 +734,7 @@ def explore(  # pylint: disable=too-many-locals,too-many-return-statements
         )
 
     @api
+    @event_logger.log_this
```
Comment (dpgaspar) [inline superset/views/core.py:737]: The order of the decorators here seem random

File: superset/views/core.py [hunk 3]
```diff
@@ -2113,6 +2133,7 @@ def stop_query(self) -> FlaskResponse:
         return self.json_response("OK")
 
     @has_access_api
+    @event_logger.log_this
```
Comment (dpgaspar) [inline superset/views/core.py:2136]: duplicate?

File: superset/views/core.py [hunk 4]
```diff
@@ -2741,6 +2766,7 @@ def sqllab(self) -> FlaskResponse:
 
     @has_access
     @expose("/sqllab/history/", methods=["GET"])
+    @event_logger.log_this
```
Comment (dpgaspar) [inline superset/views/core.py:2769]: again decorators with random order

**Review by dpgaspar — COMMENTED** [review]
Looks good, left some comments and the duplicate needs to be removed

File: superset/dashboards/api.py [hunk 5]
```diff
@@ -205,6 +206,7 @@ def __init__(self) -> None:
     @protect()
     @safe
     @statsd_metrics
+    @event_logger.log_this
```
Comment (dpgaspar) [inline superset/dashboards/api.py:209]: `log_this` depends on the DB but runs before safe, so if something fails a non handled crash will surface to the user

File: superset/views/core.py (hunk 2 above)
Comment (mistercrunch) [reply superset/views/core.py:737]: Where do you think it belongs, before `@expose`?

File: superset/datasets/api.py [hunk 6]
```diff
@@ -181,6 +181,7 @@ class DatasetRestApi(BaseSupersetModelRestApi):
     @protect()
     @safe
     @statsd_metrics
+    @event_logger.log_this
```
Comment (dpgaspar) [inline superset/datasets/api.py:184]: It's not good to have duplicate statsd metrics being sent, we can change the `log_this` decorator signature so we can pass a parameter to it (does it imply a bit refactor) or address this later

**Review by dpgaspar — COMMENTED** [review]
Looks good, left one last comment

File: superset/datasets/api.py (hunk 6 above)
Comment (mistercrunch) [reply superset/datasets/api.py]: Done!

**Comment by mistercrunch** [comment]
@dpgaspar , I think I addressed all the comments

**Review by dpgaspar — APPROVED** [review]

**Comment by ktmud** [comment]
Hmm, sorry for being late to the party. But I think `path_no_int` should've been `path_no_param` because ideally you'd want to remove all parameters (e.g. slug) even if they are not int. Also, it would be more robust to get that path from the `@expose` decorator (it adds a private `_urls[]` prop to the wrapped route handler) instead of manually cleaning up the string.

**Comment by john-bodley** [comment]
@mistercrunch, @dpgaspar, et al. would you be able to provide more as what the new fields added to the `logs` are (or will be) used for? The reason I ask is at Airbnb we use a MySQL database and the DDL operations locks the table which is problematic (especially for a large and frequently written to table like the `logs` table). Without planned downtime (possibly significant) the migration will not complete and the application is at risk of becoming inoperable (if/when a lock is obtained).

Also did you consider using the `json` column for storing the extra metadata? This could be useful especially if these fields are experimental and no DB migration would be required.

cc: @bkyryliuk

**Comment by john-bodley** [comment]
@robdiciuccio or @willbarrett do either you (or anyone at Preset) know the motivations for these changes? Wrangling this migration has already cost Airbnb quite some time and our current work around (to prevent significant downtime) is a costly and timely operation. Prior to rolling out this migration to production we just want to be sure that this change is really needed and permanent in nature, i.e., we would hate to have to perform a similar migration in the near future to rollback this logic.

**Comment by robdiciuccio** [comment]
@john-bodley looking into it and will respond shortly

**Comment by mistercrunch** [comment]
`path` seems very useful to log, and added more fields while at it, `ref` is also helpful for self-documenting the event and provide more context than the function name

Clearly this PR has 2 main parts:
1. richer logging (more fields)
1. logging more events

It sounds like db migration on `1.` is the issue for you. How many events do you have on this table? Can we help with the migration?

**Comment by mistercrunch** [comment]
@ktmud I agree about the `path_no_param` naming & approach comment, but I think identifying params may require a bit more instrumentation and work. 

About reverting (if that's on the table), we kind of need to alter the upgrade to be upgrade script to be noop AND add a downgrade script that would remove the columns if they exist as people are already on this migration.

**Comment by etr2460** [comment]
@mistercrunch we have ~190 million events in the table currently, and while we may have a solution to migrate it, I think we've historically tried to avoid major changes to the logs table since migrations can be quite costly. The last migration before yours that touched the table was in 2018, and it already includes the `json` column where we could easily add any new metadata to here.

If we do think the approach i mentioned here makes sense (putting your new fields in the json column). Then perhaps we can fix forward with a new migration, and include details on how to skip past the 2 migrations adding and removing the new columns in UPDATING.md

**Comment by john-bodley** [comment]
In general I sense we shouldn't be storing columns (or fields in JSON objects, etc.) which can easily be derived from other columns, i.e., `path_no_int` from `path`. This adds unnecessary bloat especially when the use case is not well defined. 

As I mentioned in [this](https://github.com/apache/incubator-superset/pull/11714#issuecomment-737573134) comment and as @etr2460 also stated using the `json` column seems like a better place to store experimental fields (or those subject to change) especially when they importance is secondary. For example the `json` column could include a `request` field with all the nested request properties (arguments, referrer, path, etc.). This provides a level of hierarchy/organization and flexibility.

**Comment by ktmud** [comment]
@mistercrunch I think for things as heavy as that requiring a db migration, a little bit more instrumentation is warranted. I'd vote for revert before this goes out to more users.

My two cents on the value of the added columns:

1. As John commented, `path_no_int` or `path_no_param` is not very useful because you can easily do the same RegExp matching and cleaning up in SQL.
2. The other two columns are actually under the same category as they could both have a 1:1 mapping to the existing column `action`. Since it's rare to run analytic queries on the production database because most likely you would analyze the logs in some kind of data warehouse, a little bit extra ETL work is probably more favorable in the spirit of keeping the production database lean and performant.

I agree both `path` and `ref` (maybe call this `object_ref` to disambiguate with `referrer`?) are useful information. It's also valuable to record point-at-time information even if we could manually curate the mapping. So adding them to `json` sounds like a good idea.

File: superset/views/base_api.py [hunk 7]
```diff
@@ -312,49 +313,61 @@ def info_headless(self, **kwargs: Any) -> Response:
         """
         Add statsd metrics to builtin FAB _info endpoint
         """
-        duration, response = time_function(super().info_headless, **kwargs)
-        self.send_stats_metrics(response, self.info.__name__, duration)
-        return response
+        ref = f"{self.__class__.__name__}.info"
+        with log_context(ref, ref, log_to_statsd=False):
```
Comment (ktmud) [inline superset/views/base_api.py:317]: @mistercrunch do you mind clarify why we added logs to these headless requests, as well? It seems all they does is sending statsd metrics. It doesn't make sense to me to keep a log in metadata db for when we send a log to statsd...

**Pull request merged**